pyfarm/agent/http/core/client.py

```diff
@@ -304,35 +323,56 @@ def request(method, url, **kwargs):
         used by the unittests but could be used elsewhere to add some
         custom behavior to the unpack process for the incoming response.
+
+    :raises NotImplementedError:
+        Raised whenever a request is made of this function that we can't
+        implement such as an invalid http scheme, request method or a problem
+        constructing data to an api.
     """
-    # check assumptions for arguments
-    assert method in ("HEAD", "GET", "POST", "PUT", "PATCH", "DELETE")
-    assert isinstance(url, STRING_TYPES) and url
+    assert isinstance(url, STRING_TYPES)
+
+    # Make sure it's a method we support.
+    if method not in HTTP_METHODS:
+        raise NotImplementedError(
+            "This function only supports these http "
+            "methods: %s" % ", ".join(HTTP_METHODS))
+
+    # We only support http[s]
+    parsed_url = urlparse(url)
+    if parsed_url.scheme not in HTTP_SCHEMES:
+      raise NotImplementedError("Only http or https is supported.")
+
+    # When using HTTPs there are some specific modules that are
+    # required but not specified as required in the setup.py file.
+    if parsed_url.scheme == "https":
+        if PyOpenSSL is NotImplemented and ssl is NotImplemented:
+            raise NotImplementedError(
+                "HTTPs requires the ssl or PyOpenSSL modules")
+
+        if service_identity is NotImplemented:
+            raise NotImplementedError(
+                "The service_identity module is required")
+
+    if not parsed_url.hostname:
+        raise NotImplementedError("No hostname present in url")
+
+    if not parsed_url.path:
+        raise NotImplementedError("No path provided in url")
 
     original_request = Request(
         method=method, url=url, kwargs=ImmutableDict(kwargs.copy()))
-    data = kwargs.pop("data", NOTSET)
-    headers = kwargs.pop("headers", {})
-    callback = kwargs.pop("callback", None)
-    errback = kwargs.pop("errback", log.err)
-    response_class = kwargs.pop("response_class", Response)
 
-    # check assumptions for keywords
-    assert callback is not None, "callback not provided"
-    assert callable(callback) and callable(errback)
-    assert data is NOTSET or \
-           isinstance(data, tuple(list(STRING_TYPES) + [dict, list]))
-
-    # add our default headers
+    # Headers
+    headers = kwargs.pop("headers", {})
     headers.setdefault("Content-Type", ["application/json"])
     headers.setdefault("User-Agent", [USERAGENT])
 
-    # ensure all values in the headers are lists (needed by Twisted)
+    # Twisted requires lists for header values
     for header, value in headers.items():
         if isinstance(value, STRING_TYPES):
             headers[header] = [value]
 
-        # for our purposes we should not expect headers with more
+        # For our purposes we should not expect headers with more
         # than one value for now
         elif isinstance(value, (list, tuple, set)):
             assert len(value) == 1
```

Comment:
More than one instance of a header key is technically allowed by HTTP, although unusual in practice. Why do we need to explicitly catch that case here?